main_loop_tf/utils.py

```diff
@@ -49,30 +51,53 @@ def apply_loss(labels, net_out, loss_fn, weight_decay, is_training,
 
     cfg = gflags.cfg
 
-    if mask_voids and len(cfg.void_labels):
-        # TODO Check this
-        print('Masking the void labels')
-        mask = tf.not_equal(labels, cfg.void_labels)
-        labels *= tf.cast(mask, 'int32')  # void_class --> 0 (random class)
-        # Train loss
-        loss = loss_fn(labels=labels,
-                       logits=tf.reshape(net_out, [-1, cfg.nclasses]))
-        mask = tf.cast(mask, 'float32')
-        loss *= mask
+    if cfg.of_prediction:
+        net_out, of_preds = net_out
+
+    if cfg.task in (cfg.task_names['seg'], cfg.task_names['class']):
+        if mask_voids and len(cfg.void_labels):
+            # TODO Check this
+            print('Masking the void labels')
+            mask = tf.not_equal(labels, cfg.void_labels)
+            labels *= tf.cast(mask, 'int32')  # void_class --> 0 (random class)
+            # Train loss
+            loss = loss_fn(labels=labels,
+                           logits=tf.reshape(net_out, [-1, cfg.nclasses]))
+            mask = tf.cast(mask, 'float32')
+            loss *= mask
+        else:
+            # Train loss
+            loss = loss_fn(labels=labels,
+                           logits=tf.reshape(net_out, [-1, cfg.nclasses]))
     else:
-        # Train loss
-        loss = loss_fn(labels=labels,
-                       logits=tf.reshape(net_out, [-1, cfg.nclasses]))
+        if loss_fn is tf.losses.mean_squared_error:
+            loss = loss_fn(labels=labels,
+                           predictions=tf.reshape(net_out, [-1]))
+        elif loss_fn is tf.nn.sigmoid_cross_entropy_with_logits:
+            loss = loss_fn(labels=tf.cast(labels, cfg._FLOATX),
+                           logits=tf.reshape(net_out, [-1]))
+        # TODO: else statement
 
     if is_training:
         loss = apply_l2_penalty(loss, weight_decay)
 
     # Return the mean loss (over pixels *and* batches)
     if return_mean_loss:
         if mask_voids and len(cfg.void_labels):
-            return tf.reduce_sum(loss) / tf.reduce_sum(mask)
+            loss = tf.reduce_sum(loss) / tf.reduce_sum(mask)
         else:
-            return tf.reduce_mean(loss)
+            loss = tf.reduce_mean(loss)
+        if cfg.of_regularization_type is not 'None':
```

Comment:
Move this part in a separate fork as we discussed.

Replace it with the code to compute a secondary regression loss on the optical flow (to be added to the main loss).